src/jsonrpc-virtual-machine.cpp

```diff
@@ -48,90 +57,153 @@ std::string jsonrpc_post_data(const std::string &method, const std::tuple<Ts...>
     return jsonrpc_post_data(method, params, std::make_index_sequence<sizeof...(Ts)>{});
 }
 
-struct http_request_data {
-    const std::string &url;       // NOLINT(cppcoreguidelines-avoid-const-or-ref-data-members)
-    const std::string &post_data; // NOLINT(cppcoreguidelines-avoid-const-or-ref-data-members)
-    std::string status_code;
-    std::string reason_phrase;
-    std::string entity_body;
-    bool done;
-};
-
-// Performs additional client socket configuration
-static void setup_client_socket(struct mg_connection *c) {
-#if defined(SO_LINGER)
-    // Minimize socket close time by setting the linger time to 0.
-    // On macOS, it avoids accumulating socket in TIME_WAIT state, after rapid successive requests, which can consume
-    // all available ports. It is safe to do this because it is the client who decides to close the connection, after
-    // all data is received.
-    struct linger so_linger {};
-    so_linger.l_onoff = 1;
-    so_linger.l_linger = 0; // do not wait
-    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
-    auto socket = static_cast<MG_SOCKET_TYPE>(reinterpret_cast<size_t>(c->fd));
-    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
-    (void) setsockopt(socket, SOL_SOCKET, SO_LINGER, reinterpret_cast<char *>(&so_linger), sizeof(so_linger));
-#endif
-}
-
-// Print HTTP response and signal that we're done
-static void json_post_fn(struct mg_connection *c, int ev, void *ev_data, void *fn_data) {
-    http_request_data *data = static_cast<http_request_data *>(fn_data);
-    if (ev == MG_EV_CONNECT) {
-        setup_client_socket(c);
-        const struct mg_str host = mg_url_host(data->url.c_str());
-        mg_printf(c,
-            "POST %s HTTP/1.0\r\n"
-            "Host: %.*s\r\n"
-            "Content-Type: application/json\r\n"
-            "Content-Length: %d\r\n"
-            "\r\n",
-            mg_url_uri(data->url.c_str()), static_cast<int>(host.len), host.ptr, data->post_data.size());
-        mg_send(c, data->post_data.data(), data->post_data.size());
-    } else if (ev == MG_EV_HTTP_MSG) {
-        struct mg_http_message *hm = static_cast<struct mg_http_message *>(ev_data);
-        data->entity_body = std::string_view(hm->body.ptr, hm->body.len);
-        data->status_code = std::string_view(hm->uri.ptr, hm->uri.len);
-        data->reason_phrase = std::string_view(hm->proto.ptr, hm->proto.len);
-        c->is_closing = 1;
-        data->done = true;
-    } else if (ev == MG_EV_ERROR) {
-        data->entity_body.clear();
-        data->status_code = "503";
-        data->reason_phrase = static_cast<char *>(ev_data);
-        data->done = true;
-    } else if (ev == MG_EV_CLOSE && !data->done) {
-        data->entity_body.clear();
-        data->status_code.clear();
-        data->reason_phrase = "connection closed";
-        data->done = true;
+static std::tuple<std::string, uint16_t, std::string> parse_url(const std::string &url) {
```

Comment:
Use https://www.boost.org/doc/libs/1_85_0/libs/url/doc/html/url/urls.html?

Reply:
I did this the first time, however Boost URL needs linking and my goal was to use header-only libraries, to make cross compilation easier. I think requiring to compile and link Boost libraries just for doing a small parsing is not justified. 

Boost URL used to have a header only mode before, but the author removed recently, saying that header-only libraries are bad because makes compilation slower.

Reply:
For something small like this, this is fine. But we have a running disagreement on what is enough of a reason for you to reimplement functionality that exists elsewhere. Anyway, this seems to accept a string in the form `<host>:<port>/<path-etc>` with a mandatory `:<port>`. I don 't know why `/<path-etc>` is accepted. `<scheme>:` is not accepted and neither is the `//` before the `<host>`. If that's what was intended, just add a comment that this is not a url and we are done. I am sure the previous code was worse.

Reply:
Good point, I was not happy with the URL parsing either. I removed the extra URL related code, because it's not really a URL, the old code plus Boost documentation influenced me to parse as URL, now it is simpler.

Reply:
The parsing now is a bit weird. It accepts `<host>:<port>` or just `<host>` if `<host>` is a starts with a digit. In which case it will use `<host>` as the host, and the numeric prefix in `<host>` as the port. Right? 😛

Reply:
Indeed, I've improved the parsing to check better malformed addresses.